File: health-auth/src/main/java/com/example/healthauth/configuration/WebClientConfiguration.java
package com.example.healthauth.configuration;

import io.netty.channel.ChannelOption;
import io.netty.handler.logging.LogLevel;
import io.netty.handler.timeout.ReadTimeoutHandler;
import io.netty.handler.timeout.WriteTimeoutHandler;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;
import org.springframework.http.client.reactive.ReactorClientHttpConnector;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.netty.http.client.HttpClient;
import reactor.netty.transport.logging.AdvancedByteBufFormat;

@Configuration
public class WebClientConfiguration {

    @Value("${keycloak.base-url}")
    private String keycloakBaseUrl;

    @Bean
    @Primary
    public WebClient webclient() {
        return WebClient.builder()
            .clientConnector(
                new ReactorClientHttpConnector(
                    HttpClient.create()
                        .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 15_000)
                        .doOnConnected( (it) -> {
                            it.addHandlerLast(new ReadTimeoutHandler(8));
                            it.addHandlerLast(new WriteTimeoutHandler(8));
                            })
                        .wiretap("reactor.netty.http.client.HttpClient",
                            LogLevel.INFO, AdvancedByteBufFormat.TEXTUAL)
                        .compress(false)
                )
            )
            .baseUrl(keycloakBaseUrl)
            .build();
    }
}

File: health-gateway/src/main/kotlin/com/example/heathgateway/config/WebclientConfiguration.kt
package com.example.heathgateway.config

import io.netty.channel.ChannelOption
import io.netty.handler.timeout.ReadTimeoutHandler
import io.netty.handler.timeout.WriteTimeoutHandler
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.client.reactive.ReactorClientHttpConnector
import org.springframework.http.codec.ClientCodecConfigurer
import org.springframework.http.codec.HttpMessageWriter
import org.springframework.http.codec.LoggingCodecSupport
import org.springframework.web.reactive.function.client.ExchangeStrategies
import org.springframework.web.reactive.function.client.WebClient
import reactor.netty.http.client.HttpClient
import reactor.netty.resources.ConnectionProvider
import java.time.Duration

@Configuration
class WebclientConfiguration {

    @Bean
    fun defaultWebclient(): WebClient {

        val exchangeStrategies = ExchangeStrategies.builder()
            .codecs { configurer: ClientCodecConfigurer ->
                configurer.defaultCodecs().maxInMemorySize(1024 * 1024 * 50)
            }
            .build()

        exchangeStrategies
            .messageWriters().stream()
            .filter { obj: HttpMessageWriter<*>? ->
                LoggingCodecSupport::class.java.isInstance(
                    obj
                )
            }
            .forEach { writer: HttpMessageWriter<*> ->
                (writer as LoggingCodecSupport).isEnableLoggingRequestDetails = true
            }

        val provider = ConnectionProvider.builder("webclient-connection-pool")
            .maxConnections(500)
            .pendingAcquireTimeout(Duration.ofMillis(0))
            .pendingAcquireMaxCount(-1)
            .maxIdleTime(Duration.ofMillis(8000L))
            .maxLifeTime(Duration.ofMillis(8000L))
            .build()

        return WebClient.builder()
            .clientConnector(
                ReactorClientHttpConnector(
                    HttpClient.create(provider)
                        .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 15_000)
                        .doOnConnected {
                            it.addHandlerLast(ReadTimeoutHandler(8))
                            it.addHandlerLast(WriteTimeoutHandler(8))
                        }
                        .wiretap(true)
                )
            )
            .exchangeStrategies(exchangeStrategies)
            .build()
    }
}
